validator-api/src/cache/mod.rs

```diff
@@ -214,31 +216,87 @@ impl ValidatorCache {
 
-        // generate list of mixnodes and their relatively weight (by total stake)
-        let choices = mixnodes
-            .iter()
-            .map(|mix| {
-                // note that the theoretical maximum possible stake is equal to the total
-                // supply of all tokens, i.e. 1B (which is 1 quadrillion of native tokens, i.e. 10^15 ~ 2^50)
-                // which is way below maximum value of f64, so the cast is fine
-                let total_stake = mix.total_stake().unwrap_or_default() as f64;
-                (mix, total_stake)
-            }) // if for some reason node is invalid, treat it as 0 stake/weight
-            .collect::<Vec<_>>();
+        self.stake_weighted_choice(mixnodes, nodes_to_select as usize, &mut rng)
+    }
 
+    fn stake_weighted_choice(
+        &self,
+        mixnodes: &[MixNodeBond],
+        nodes_to_select: usize,
+        rng: &mut ChaCha20Rng,
+    ) -> Vec<MixNodeBond> {
+        // generate list of mixnodes and their relatively weight (by total stake)
+        let choices = self.generate_mixnode_stake_tuples(mixnodes);
         // the unwrap here is fine as an error can only be thrown under one of the following conditions:
         // - our mixnode list is empty - we have already checked for that
         // - we have invalid weights, i.e. less than zero or NaNs - it shouldn't happen in our case as we safely cast down from u128
         // - all weights are zero - it's impossible in our case as the list of nodes is not empty and weight is proportional to stake. You must have non-zero stake in order to bond
         // - we have more than u32::MAX values (which is incredibly unrealistic to have 4B mixnodes bonded... literally every other person on the planet would need one)
         choices
-            .choose_multiple_weighted(&mut rng, nodes_to_select as usize, |item| item.1)
+            .choose_multiple_weighted(rng, nodes_to_select, |item| item.1)
             .unwrap()
-            .map(|(bond, _weight)| *bond)
+            .map(|(bond, _weight)| bond)
             .cloned()
             .collect()
     }
 
+    fn generate_mixnode_stake_tuples(&self, mixnodes: &[MixNodeBond]) -> Vec<(MixNodeBond, f64)> {
+        mixnodes
+            .iter()
+            .map(|mix| {
+                // note that the theoretical maximum possible stake is equal to the total
+                // supply of all tokens, i.e. 1B (which is 1 quadrillion of native tokens, i.e. 10^15 ~ 2^50)
+                // which is way below maximum value of f64, so the cast is fine
+                let total_stake = mix.total_stake().unwrap_or_default() as f64;
+                (mix.clone(), total_stake)
+            }) // if for some reason node is invalid, treat it as 0 stake/weight
+            .collect()
+    }
+
+    // Estimate probability that a node will end up in the rewarded set, by running the selection process 100 times and aggregating the results.
```

Comment:
Why should we run the selection process N times and then aggregate the result? The node's probability is just a ratio of its total bond over all bonded tokens and with probability being 1 if number of bonded nodes <= K
My point being, if the network doesn't change, the probability should always be the same for given node

Reply:
Aha. Found something possibly useful: https://math.stackexchange.com/questions/316175/probability-to-choose-specific-item-in-a-weighted-sampling-without-replacement

Reply:
+1 to that, I don't think we should simulate it

Reply:
This is an excellent point, I think I was just being lazy and not wanting to do the math for active vs reserve set, although its not much of a math, its just stacking probabilities, let me revisit that

Reply:
Though it kinda seems like a possibly expensive computation. Since it doesn't really depend on validator's state, perhaps it might be worth to compute it client's side so that validator-api couldn't be overwhelmed with those?

Reply:
I'm hoping Rust crush those calculations, I guess it depends on how it's used, if the wallet will only make one request at a time I think we'll be fine, but worth testing for sure

Reply:
Ok, going with straightforward probability calculations, we're approximating some, but not a lot.